server/fingerprint.go

```diff
@@ -0,0 +1,113 @@
+package server
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+)
+
+type Fingerprint uint64
+
+// FingerprintFromRequest returns a fingerprint for the request based on the X-Forwarded-For header
+// and a salted timestamp.  The fingerprint is used to identify unique users sessions
+// over a short period of time, and thus can be used as a key for rate limiting.
+func FingerprintFromRequest(req *http.Request, at time.Time) (Fingerprint, error) {
+	// X-Forwarded-For header contains a comma-separated list of IP addresses when
+	// the request has been forwarded through multiple proxies.  For example:
+	//
+	// X-Forwarded-For: 2600:8802:4700:bee:d13c:c7fb:8e0f:84ff, 172.70.210.100
+	xff, err := getXForwardedForIP(req)
+	if err != nil {
+		return 0, err
+	}
+	// We considered adding the User-Agent header to the fingerprint, but decided
```

Comment:
this could be part of the overall function comment as it's important and general